src/toil/server/wes/runner.py

```diff
@@ -0,0 +1,327 @@
+from typing import Dict, Any, List, Union
+
+from toil.server.utils import get_iso_time, get_file_class, link_file
+from toil.statsAndLogging import configure_root_logger, set_log_level
+
+logger = logging.getLogger(__name__)
+
+
+"""
+Toil WES workflow runner.
+
+Helper script to run a requested WES workflow through subprocess. Responsible
+for parsing the user request into a shell command, executing that command, and
+collecting the outputs of the resulting workflow run.
+This script expects the given `--work_dir` directory to contain an `execution`
+directory where workflow attachments are staged, and a `request.json` JSON file
+containing the body of the request:
+
+.
+├── execution/
+│   └── ...
+└── request.json
+
+During execution, this script may write the following files to the working
+directory:
+
+.
+├── cmd
+├── end_time
+├── exit_code
+├── job_store
+├── outputs.json
+├── pid
+├── start_time
+├── state
+├── stderr
+└── stdout
+"""
+
+
+class ToilWorkflowRunner:
+    """
+    A class to represent a workflow runner to run the requested workflow.
+
+    Responsible for parsing the user request into a shell command, executing
+    that command, and collecting the outputs of the resulting workflow run.
+    """
+
+    def __init__(self, work_dir: str, request: Dict[str, Any], engine_options: List[str]):
+        self.work_dir = work_dir
+        self.request = request
+        self.engine_options = engine_options
+
+        self.wf_type: str = request["workflow_type"].lower().strip()
+        self.version: str = request["workflow_type_version"]
+
+        self.exec_dir = os.path.join(self.work_dir, "execution")
+        self.out_dir = os.path.join(self.work_dir, "outputs")
+
+        self.default_job_store = "file:" + os.path.join(self.work_dir, "toil_job_store")
+        self.job_store = self.default_job_store
+
+    def write(self, filename: str, contents: str) -> None:
+        with open(os.path.join(self.work_dir, filename), "w") as f:
+            f.write(contents)
+
+    def get_state(self) -> str:
+        with open(os.path.join(self.work_dir, "state"), "r") as f:
+            return f.read()
+
+    def set_state(self, state: str) -> None:
+        self.write("state", state)
+
+    def write_workflow(self, src_url: str) -> str:
+        """
+        Fetch the workflow file from its source and write it to a destination
+        file. The destination is returned.
+        """
+        logger.info(f"Processing workflow_url: '{src_url}'...")
+        dest = src_url
+
+        if src_url.startswith("file://"):
+            logger.info(f"Linking workflow from filesystem.")
+            dest = os.path.join(self.exec_dir, os.path.basename(src_url))
+            link_file(src=src_url[7:], dest=dest)
+        elif src_url.startswith(("http://", "https://")):
+            logger.info(f"Downloading workflow_url from the Internet.")
+            # TODO: Download workflow files from the Internet
+            raise NotImplementedError
```

Comment:
This probably be just a simple `requests.get()` call with a check to make sure it is, in fact, the contents of a workflow and not a dancing penguin gif.